### PR TITLE
Alias for Lookup with Relationship

### PR DESCRIPTION
**Pull request by josemiguelq**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | 

#### Summary

According to the documentation of [Mongo](https://docs.mongodb.com/manual/reference/operator/aggregation/lookup/) the `as` can't be null because contains the matching documents from the from collection.
(https://docs.mongodb.com/manual/reference/operator/aggregation/lookup/#lookup-join-as).

To prevent error the keys will be the same of the Documents using "mappedBy" and the foreignKey will be the same of attribute "name".

### CODE REVIEW

**Comment by alcaeus** [comment]
Thanks for your contribution! This needs a little more work:
1. Please give your commit a good title. "Use field name as default alias in Lookup stage" would be a good title.
2. Rebase your commit on the 1.3.x branch and re-target the pull request to that branch, as this is a bug we should fix in the oldest supported version
3. Please add a test for it, there are a couple of tests for the Lookup stage already.

Please let me know if you need help with any of the above!

**Comment by josemiguelq** [comment]
> Thanks for your contribution! This needs a little more work:
> 
> 1. Please give your commit a good title. "Use field name as default alias in Lookup stage" would be a good title.
> 2. Rebase your commit on the 1.3.x branch and re-target the pull request to that branch, as this is a bug we should fix in the oldest supported version
> 3. Please add a test for it, there are a couple of tests for the Lookup stage already.
> 
> Please let me know if you need help with any of the above!

Thanks @alcaeus , I'll do the corrections that you pointed.

**Comment by josemiguelq** [comment]
@alcaeus , checks my updates please?

File: tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php [hunk 1]
```diff
@@ -420,4 +420,50 @@ private function insertTestData()
...
+
+        $result = $builder->execute()->toArray();
+        $this->assertCount(1, $result[0]['simpleReferenceOneInverse']);
+    }
+
+    public function testLookupStageAndDefaultAliasOverride()
+    {
+        $builder = $this->dm->createAggregationBuilder(\Documents\User::class);
+        $builder
+            ->lookup('simpleReferenceOneInverse')->alias('override');
```
Comment (alcaeus) [inline tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php]: Please fix indentation:
```suggestion
            ->lookup('simpleReferenceOneInverse')
                ->alias('override');
```

File: tests/Documents/ReferenceCampaign.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace Documents;
+
+use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
+
+/** @ODM\Document */
+class ReferenceCampaign
```
Comment (alcaeus) [inline tests/Documents/ReferenceCampaign.php]: This class is not used and can be removed.

File: tests/Documents/Campaign.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Documents;
+
+use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
+
+/**
+ * @ODM\Document(collection="campaigns")
+ * @ODM\InheritanceType("COLLECTION_PER_CLASS")
+ */
+class Campaign
```
Comment (alcaeus) [inline tests/Documents/Campaign.php]: This class is not used and can be removed.

File: tests/Documents/User.php [hunk 4]
```diff
@@ -74,7 +74,7 @@ class User extends BaseDocument
     /** @ODM\ReferenceMany(targetDocument="BlogPost", mappedBy="user", nullable=true) */
     protected $posts;
 
-    /** @ODM\ReferenceOne(targetDocument="Documents\SimpleReferenceUser", mappedBy="user") */
+    /** @ODM\ReferenceOne(targetDocument="Documents\SimpleReferenceUser", mappedBy="user", name="userId",storeAs="id") */
```
Comment (alcaeus) [inline tests/Documents/User.php]: Why are you changing the mapping for this? `storeAs` should not be necessary as inverse references don't store the identifier anyways.
The `name` attribute should also be removed, you should use the actual name (`simpleReferenceOneInverse`) in the test instead.

File: tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php [hunk 5]
```diff
@@ -47,9 +47,9 @@ public function testLookupStageWithFieldNameTranslation()
         $builder = $this->dm->createAggregationBuilder(\Documents\SimpleReferenceUser::class);
         $builder
             ->lookup(CmsComment::class)
-                ->localField('id')
```
[the remark this answers is not in the document]
Comment (josemiguelq) [reply tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php]: done

File: tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php [hunk 6]
```diff
@@ -70,9 +70,9 @@ public function testLookupStageWithClassName()
         $builder = $this->dm->createAggregationBuilder(\Documents\SimpleReferenceUser::class);
         $builder
             ->lookup(\Documents\User::class)
-                ->localField('userId')
```
[the remark this answers is not in the document]
Comment (josemiguelq) [reply tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php]: done

File: tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (josemiguelq) [reply tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php]: done

File: tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php [hunk 7]
```diff
@@ -381,8 +381,8 @@ public function testLookupToShardedCollectionThrowsException()
         $this->expectException(MappingException::class);
         $builder
             ->lookup(ShardedOne::class)
-                ->localField('id')
```
[the remark this answers is not in the document]
Comment (josemiguelq) [reply tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php]: done

File: tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php [hunk 8]
```diff
@@ -128,7 +128,7 @@ public function testLookupStageReferenceMany()
         $builder
             ->unwind('$users')
             ->lookup('users')
-                ->alias('users');
```
[the remark this answers is not in the document]
Comment (josemiguelq) [reply tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php]: done

File: tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php [hunk 9]
```diff
@@ -99,9 +99,9 @@ public function testLookupStageWithCollectionName()
         $builder = $this->dm->createAggregationBuilder(\Documents\SimpleReferenceUser::class);
         $builder
             ->lookup('randomCollectionName')
-                ->localField('userId')
```
[the remark this answers is not in the document]
Comment (josemiguelq) [reply tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/LookupTest.php]: done

File: tests/Documents/User.php (hunk 4 above)
Comment (josemiguelq) [reply tests/Documents/User.php]: sure, my bad

**Comment by josemiguelq** [comment]
@alcaeus , request changes done. Thanks for the review and advices

**Comment by alcaeus** [comment]
@josemiguelq there are some more indentation changes that you haven't changed back. Please go through those and make sure to not have any unrelated changes in. The rest looks good 👍

**Comment by josemiguelq** [comment]
> @josemiguelq there are some more indentation changes that you haven't changed back. Please go through those and make sure to not have any unrelated changes in. The rest looks good 👍

Sorry @alcaeus, indentation changes done. Its IDE default behavior

**Comment by alcaeus** [comment]
Thanks @josemiguelq!

Note: travis-ci completed successfully but did not correctly report the status to GitHub.

**Pull request merged**